app/components/graph-editor/drawable-interfaces.ts

```diff
@@ -0,0 +1,398 @@
+    clone.borderWidth = n.borderWidth;
+    return clone;
+}
+
+
+// Interfaces //////////////////////////////////////////////////////////////////
+
+
+/**
+ * GraphContext  
+ *   Interface that exposes a graph and any external properties that the 
+ *   editor needs to have to go with it. Specifically the set of selected
+ *   `DrawableNode | DrawableEdge`
+ */
+export interface GraphContext {
+    readonly graph: DrawableGraph;
+    readonly selectedDrawables: Set<Drawable>;
+}
+
+/**
+ * DrawableGraph  
+ *   Interface that exposes drawable graph properties and methods.
+ */
+export interface DrawableGraph {
+
+    /**
+     * nodes  
+     *   The collection of drawable nodes that are part of the graph.
+     */
+    readonly nodes: Iterable<DrawableNode>;
+
+    /**
+     * edges  
+     *   The collection of drawable edges that are part of the graph.
+     */
+    readonly edges: Iterable<DrawableEdge>;
+
+    /**
+     * backgroundColor  
+     *   What is this doing here? It should be a public property of the graph
+     *   editor component.
+     */
+    backgroundColor: string;
+
+    /**
+     * createNode  
+     *   Creates a drawable node with an optional position.
+     */
+    createNode(x?: number, y?: number): DrawableNode;
+
+    /**
+     * removeNode  
+     *   Guarantees that the given node is not present in the graph.
+     */
+    removeNode(node: DrawableNode): void;
+
+    /**
+     * canCreateEdge  
+     *   Checks if a drawable edge can be created with the given source and
+     *   destination nodes.
+     * 
+     *   If `like` is specified, a drawable edge with a matching type of `like` is
+     *   is checked against the given source and destination nodes.
+     * 
+     *   This method must be called before calling `createEdge`.
+     */
+    canCreateEdge(
+        src: DrawableNode,
+        dst: DrawableNode,
+        like?: DrawableEdge
+    ): boolean;
+
+    /**
+     * createEdge  
+     *   Creates a drawable edge with a source and destination node.
+     * 
+     *   If `like` is specified, a drawable edge with a matching type of `like` is
+     *   created.
+     * 
+     *   The `canCreateEdge` method must be called to check if creating the edge
+     *   is valid.
+     */
+    createEdge(
+        src: DrawableNode,
+        dst: DrawableNode,
+        like?: DrawableEdge
+    ): DrawableEdge;
+
+    /**
+     * replaceEdge  
+     *   This method has been replaced by `createEdge` and is no longer needed.
+     */
+    replaceEdge(original: DrawableEdge, replacement: DrawableEdge): void;
+
+    /**
+     * removeEdge  
+     *   Guarantees that the given edge is not present in the graph.
+     */
+    removeEdge(edge: DrawableEdge): void;
+
+}
+
+/**
+ * DrawableEdge  
+ *   Interface that exposes drawable edge properties.
+ */
+export interface DrawableEdge {
+
+    /**
+     * source  
+     *   The source node of the edge.
+     */
+    source: DrawableNode | null;
```

Comment:
I may change this to be a definite `DrawableNode`, since it'll be easier to not draw hidden graph elements with the code refactor in place. Originally, the `| null` was so that I can have one end of the edge follow the cursor as it is being dragged in the case of moving an edge from one node to another.